hash/hash.go

```diff
@@ -12,34 +14,66 @@ const LabelHashNestingDepth = 100
 // concatenation of the left- and right-child in the tree and the result is hashed using Sha256.
 //
 // ⚠️ The resulting function is NOT thread-safe, however different generated instances are independent.
-// The code is optimized for performance and memory allocations.
 func GenMerkleHashFunc(challenge []byte) merkle.HashFunc {
-	h := sha256.New()
-	return func(buf, lChild, rChild []byte) []byte {
-		h.Reset()
-		_, _ = h.Write(challenge)
-		_, _ = h.Write(lChild)
-		_, _ = h.Write(rChild)
-		return h.Sum(buf)
+	// De-virtualize the call to sha256 hasher
+	switch h := sha256.New().(type) {
+	case *sha256.Digest:
+		return func(buf, lChild, rChild []byte) []byte {
+			if cap(buf) < sha256.Size {
+				buf = make([]byte, sha256.Size)
+			} else {
+				buf = buf[:sha256.Size]
+			}
+			h.Reset()
+			_, _ = h.Write(challenge)
+			_, _ = h.Write(lChild)
+			_, _ = h.Write(rChild)
+			h.CheckSumInto((*[sha256.Size]byte)(buf))
+			return buf
+		}
+	default:
+		return func(buf, lChild, rChild []byte) []byte {
+			h.Reset()
+			_, _ = h.Write(challenge)
+			_, _ = h.Write(lChild)
+			_, _ = h.Write(rChild)
+			return h.Sum(buf)
```

Comment:
I don't think this is correct, this should probably be

```suggestion
			return h.Sum(buf[:0])
```

otherwise the content of `buf` will be used as input to the hash.